prometheus-to-sd/translator/metric_descriptor_cache.go

```diff
@@ -0,0 +1,174 @@
+	// of view, we don't want to check all metric descriptors too often, as they should change rarely.
+	if !cache.fresh {
+		return
+	}
+	for _, metricFamily := range metrics {
+		if metricWhitelisted(metricFamily.GetName(), whitelisted) {
+			cache.updateMetricDescriptorIfStale(metricFamily)
+		}
+	}
+}
+
+func metricWhitelisted(metric string, whitelisted []string) bool {
+	// Empty list means that we want to fetch all metrics.
+	if len(whitelisted) == 0 {
+		return true
+	}
+	for _, whitelistedMetric := range whitelisted {
+		if whitelistedMetric == metric {
+			return true
+		}
+	}
+	return false
+}
+
+// updateMetricDescriptorIfStale checks if descriptor created from MetricFamily object differs from the existing one
+// and updates if needed.
+func (cache *MetricDescriptorCache) updateMetricDescriptorIfStale(metricFamily *dto.MetricFamily) {
+	metricDescriptor, ok := cache.descriptors[metricFamily.GetName()]
+	updatedMetricDescriptor := MetricFamilyToMetricDescriptor(cache.config, metricFamily, metricDescriptor)
+	if strings.HasPrefix(updatedMetricDescriptor.Type, customMetricsPrefix) &&
+		(!ok || descriptorChanged(metricDescriptor, updatedMetricDescriptor)) {
+		cache.updateMetricDescriptorInStackdriver(updatedMetricDescriptor)
+		cache.descriptors[metricFamily.GetName()] = updatedMetricDescriptor
+	}
+}
+
+func descriptorChanged(original *v3.MetricDescriptor, checked *v3.MetricDescriptor) bool {
+	if original.Description != checked.Description {
+		glog.V(4).Infof("Description is different, %v != %v", original.Description, checked.Description)
+		return true
+	}
+	for _, label := range checked.Labels {
+		found := false
+		for _, labelFromOriginal := range original.Labels {
+			if label.Key == labelFromOriginal.Key {
+				found = true
+				break
+			}
+		}
+		if !found {
+			glog.V(4).Infof("Missing label %v in the original metric descriptor", label)
+			return true
+		}
+	}
+	return false
+}
+
+// updateMetricDescriptorInStackdriver writes metric descriptor to the stackdriver.
+func (cache *MetricDescriptorCache) updateMetricDescriptorInStackdriver(metricDescriptor *v3.MetricDescriptor) {
+	glog.V(4).Infof("Updating metric descriptor: %+v", metricDescriptor)
+
+	projectName := createProjectName(cache.config.GceConfig)
+	_, err := cache.service.Projects.MetricDescriptors.Create(projectName, metricDescriptor).Do()
+	if err != nil {
+		if _, metricName, err := parseMetricType(cache.config.GceConfig, metricDescriptor.Type); err == nil {
+			cache.broken[metricName] = true
+		} else {
+			glog.Warningf("Unable to parse %v: %v", metricDescriptor.Type, err)
+		}
+		glog.Errorf("Error in attempt to update metric descriptor %v", err)
+	}
+}
+
+// Refresh function fetches all metric descriptors of all metrics defined for given component with a defined prefix
+// and puts them into cache.
+func (cache *MetricDescriptorCache) Refresh() {
+	proj := createProjectName(cache.config.GceConfig)
+	cache.descriptors = make(map[string]*v3.MetricDescriptor)
```

Comment:
can we override descriptors only when the new ones are correctly fetched from SD? I think it's better to have old ones than nothing.